From 7d61f049de70934848cc3244b3e5f19be49a52a4 Mon Sep 17 00:00:00 2001
From: Jason Dere <jdere@hortonworks.com>
Date: Tue, 12 Jan 2016 17:39:05 -0800
Subject: [PATCH 1219/1363] CDH-55703: HIVE-12772: Beeline/JDBC output of
 decimal values is not 0-padded, does not match
 with CLI output (Jason Dere, reviewed by Xuefu
 Zhang)

(cherry picked from commit 9f1ff4f17525b8db5c63c7bcfb43ee73f42836eb)

Change-Id: I7c3d5ed0c7dba09321c134401d62cdde2f3f52be
---
 data/files/datatypes.txt                           |    2 +-
 .../java/org/apache/hive/jdbc/TestJdbcDriver2.java |    2 +-
 .../java/org/apache/hive/service/cli/Column.java   |   10 ++++++++++
 .../apache/hive/service/cli/ColumnBasedSet.java    |   14 +++++++-------
 .../org/apache/hive/service/cli/ColumnValue.java   |   11 +++++++----
 .../org/apache/hive/service/cli/RowBasedSet.java   |   14 +++++++-------
 .../org/apache/hive/service/cli/TableSchema.java   |    6 +++---
 7 files changed, 36 insertions(+), 23 deletions(-)

diff --git a/data/files/datatypes.txt b/data/files/datatypes.txt
index 458c5bd..0872a1f 100644
--- a/data/files/datatypes.txt
+++ b/data/files/datatypes.txt
@@ -1,3 +1,3 @@
 \N\N\N\N\N\N\N\N\N\N\N\N\N\N\N\N\N\N\N\N\N\N\N
 -1false-1.1\N\N\N-1-1-1.0-1\N\N\N\N\N\N\N\N\N
-1true1.11121x2ykva92.2111.01abcd1111213142212212x1abcd22012-04-22 09:00:00.123456789123456789.0123456YWJjZA==2013-01-01abc123abc123X'01FF'
+1true1.11121x2ykva92.2111.01abcd1111213142212212x1abcd22012-04-22 09:00:00.123456789123456789.123456YWJjZA==2013-01-01abc123abc123X'01FF'
diff --git a/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestJdbcDriver2.java b/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestJdbcDriver2.java
index 7e900c4..77906cc 100644
--- a/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestJdbcDriver2.java
+++ b/itests/hive-unit/src/test/java/org/apache/hive/jdbc/TestJdbcDriver2.java
@@ -801,7 +801,7 @@ public void testDataTypes() throws Exception {
     assertEquals("[{\"m\":{},\"n\":1},{\"m\":{\"a\":\"b\",\"c\":\"d\"},\"n\":2}]", res.getString(16));
     assertEquals("2012-04-22 09:00:00.123456789", res.getString(17));
     assertEquals("2012-04-22 09:00:00.123456789", res.getTimestamp(17).toString());
-    assertEquals("123456789.0123456", res.getBigDecimal(18).toString());
+    assertEquals("123456789.1234560", res.getBigDecimal(18).toString());
     assertEquals("abcd", res.getString(19));
     assertEquals("2013-01-01", res.getString(20));
     assertEquals("2013-01-01", res.getDate(20).toString());
diff --git a/service/src/java/org/apache/hive/service/cli/Column.java b/service/src/java/org/apache/hive/service/cli/Column.java
index 2e21f18..1c9cb77 100644
--- a/service/src/java/org/apache/hive/service/cli/Column.java
+++ b/service/src/java/org/apache/hive/service/cli/Column.java
@@ -41,6 +41,8 @@
 import org.apache.hive.service.cli.thrift.TI64Column;
 import org.apache.hive.service.cli.thrift.TStringColumn;
 
+import org.apache.hadoop.hive.common.type.HiveDecimal;
+
 /**
  * Column.
  */
@@ -325,6 +327,14 @@ public TColumn toTColumn() {
   private static final ByteBuffer EMPTY_BINARY = ByteBuffer.allocate(0);
   private static final String EMPTY_STRING = "";
 
+  public void addValue(TypeDescriptor typeDescriptor, Object field) {
+    if (field != null && typeDescriptor.getType() == Type.DECIMAL_TYPE) {
+      int scale = typeDescriptor.getDecimalDigits();
+      field = ((HiveDecimal) field).toFormatString(scale);
+    }
+    addValue(typeDescriptor.getType(), field);
+  }
+
   public void addValue(Type type, Object field) {
     switch (type) {
       case BOOLEAN_TYPE:
diff --git a/service/src/java/org/apache/hive/service/cli/ColumnBasedSet.java b/service/src/java/org/apache/hive/service/cli/ColumnBasedSet.java
index 47a582e..36a2a80 100644
--- a/service/src/java/org/apache/hive/service/cli/ColumnBasedSet.java
+++ b/service/src/java/org/apache/hive/service/cli/ColumnBasedSet.java
@@ -33,11 +33,11 @@
 
   private long startOffset;
 
-  private final Type[] types; // non-null only for writing (server-side)
+  private final TypeDescriptor[] descriptors; // non-null only for writing (server-side)
   private final List<Column> columns;
 
   public ColumnBasedSet(TableSchema schema) {
-    types = schema.toTypes();
+    descriptors = schema.toTypeDescriptors();
     columns = new ArrayList<Column>();
     for (ColumnDescriptor colDesc : schema.getColumnDescriptors()) {
       columns.add(new Column(colDesc.getType()));
@@ -45,7 +45,7 @@ public ColumnBasedSet(TableSchema schema) {
   }
 
   public ColumnBasedSet(TRowSet tRowSet) {
-    types = null;
+    descriptors = null;
     columns = new ArrayList<Column>();
     for (TColumn tvalue : tRowSet.getColumns()) {
       columns.add(new Column(tvalue));
@@ -53,8 +53,8 @@ public ColumnBasedSet(TRowSet tRowSet) {
     startOffset = tRowSet.getStartRowOffset();
   }
 
-  private ColumnBasedSet(Type[] types, List<Column> columns, long startOffset) {
-    this.types = types;
+  private ColumnBasedSet(TypeDescriptor[] descriptors, List<Column> columns, long startOffset) {
+    this.descriptors = descriptors;
     this.columns = columns;
     this.startOffset = startOffset;
   }
@@ -62,7 +62,7 @@ private ColumnBasedSet(Type[] types, List<Column> columns, long startOffset) {
   @Override
   public ColumnBasedSet addRow(Object[] fields) {
     for (int i = 0; i < fields.length; i++) {
-      columns.get(i).addValue(types[i], fields[i]);
+      columns.get(i).addValue(descriptors[i], fields[i]);
     }
     return this;
   }
@@ -89,7 +89,7 @@ public ColumnBasedSet extractSubset(int maxRows) {
     for (int i = 0; i < columns.size(); i++) {
       subset.add(columns.get(i).extractSubset(0, numRows));
     }
-    ColumnBasedSet result = new ColumnBasedSet(types, subset, startOffset);
+    ColumnBasedSet result = new ColumnBasedSet(descriptors, subset, startOffset);
     startOffset += numRows;
     return result;
   }
diff --git a/service/src/java/org/apache/hive/service/cli/ColumnValue.java b/service/src/java/org/apache/hive/service/cli/ColumnValue.java
index d383180..e25ddbf 100644
--- a/service/src/java/org/apache/hive/service/cli/ColumnValue.java
+++ b/service/src/java/org/apache/hive/service/cli/ColumnValue.java
@@ -136,15 +136,18 @@ private static TColumnValue timestampValue(Timestamp value) {
     return TColumnValue.stringVal(tStringValue);
   }
 
-  private static TColumnValue stringValue(HiveDecimal value) {
+  private static TColumnValue stringValue(HiveDecimal value, TypeDescriptor typeDescriptor) {
     TStringValue tStrValue = new TStringValue();
     if (value != null) {
-      tStrValue.setValue(value.toString());
+      int scale = typeDescriptor.getDecimalDigits();
+      tStrValue.setValue(value.toFormatString(scale));
     }
     return TColumnValue.stringVal(tStrValue);
   }
 
-  public static TColumnValue toTColumnValue(Type type, Object value) {
+  public static TColumnValue toTColumnValue(TypeDescriptor typeDescriptor, Object value) {
+    Type type = typeDescriptor.getType();
+
     switch (type) {
     case BOOLEAN_TYPE:
       return booleanValue((Boolean)value);
@@ -171,7 +174,7 @@ public static TColumnValue toTColumnValue(Type type, Object value) {
     case TIMESTAMP_TYPE:
       return timestampValue((Timestamp)value);
     case DECIMAL_TYPE:
-      return stringValue(((HiveDecimal)value));
+      return stringValue((HiveDecimal)value, typeDescriptor);
     case BINARY_TYPE:
       return stringValue((String)value);
     case ARRAY_TYPE:
diff --git a/service/src/java/org/apache/hive/service/cli/RowBasedSet.java b/service/src/java/org/apache/hive/service/cli/RowBasedSet.java
index a0ee210..6fbb4c5 100644
--- a/service/src/java/org/apache/hive/service/cli/RowBasedSet.java
+++ b/service/src/java/org/apache/hive/service/cli/RowBasedSet.java
@@ -33,22 +33,22 @@
 
   private long startOffset;
 
-  private final Type[] types; // non-null only for writing (server-side)
+  private final TypeDescriptor[] descriptors; // non-null only for writing (server-side)
   private final RemovableList<TRow> rows;
 
   public RowBasedSet(TableSchema schema) {
-    types = schema.toTypes();
+    descriptors = schema.toTypeDescriptors();
     rows = new RemovableList<TRow>();
   }
 
   public RowBasedSet(TRowSet tRowSet) {
-    types = null;
+    descriptors = null;
     rows = new RemovableList<TRow>(tRowSet.getRows());
     startOffset = tRowSet.getStartRowOffset();
   }
 
-  private RowBasedSet(Type[] types, List<TRow> rows, long startOffset) {
-    this.types = types;
+  private RowBasedSet(TypeDescriptor[] descriptors, List<TRow> rows, long startOffset) {
+    this.descriptors = descriptors;
     this.rows = new RemovableList<TRow>(rows);
     this.startOffset = startOffset;
   }
@@ -57,7 +57,7 @@ private RowBasedSet(Type[] types, List<TRow> rows, long startOffset) {
   public RowBasedSet addRow(Object[] fields) {
     TRow tRow = new TRow();
     for (int i = 0; i < fields.length; i++) {
-      tRow.addToColVals(ColumnValue.toTColumnValue(types[i], fields[i]));
+      tRow.addToColVals(ColumnValue.toTColumnValue(descriptors[i], fields[i]));
     }
     rows.add(tRow);
     return this;
@@ -75,7 +75,7 @@ public int numRows() {
 
   public RowBasedSet extractSubset(int maxRows) {
     int numRows = Math.min(numRows(), maxRows);
-    RowBasedSet result = new RowBasedSet(types, rows.subList(0, numRows), startOffset);
+    RowBasedSet result = new RowBasedSet(descriptors, rows.subList(0, numRows), startOffset);
     rows.removeRange(0, numRows);
     startOffset += numRows;
     return result;
diff --git a/service/src/java/org/apache/hive/service/cli/TableSchema.java b/service/src/java/org/apache/hive/service/cli/TableSchema.java
index ee019bc..d052268 100644
--- a/service/src/java/org/apache/hive/service/cli/TableSchema.java
+++ b/service/src/java/org/apache/hive/service/cli/TableSchema.java
@@ -82,10 +82,10 @@ public TTableSchema toTTableSchema() {
     return tTableSchema;
   }
 
-  public Type[] toTypes() {
-    Type[] types = new Type[columns.size()];
+  public TypeDescriptor[] toTypeDescriptors() {
+    TypeDescriptor[] types = new TypeDescriptor[columns.size()];
     for (int i = 0; i < types.length; i++) {
-      types[i] = columns.get(i).getType();
+      types[i] = columns.get(i).getTypeDescriptor();
     }
     return types;
   }
-- 
1.7.9.5

